paasta_tools/cli/cmds/exec.py

```diff
@@ -0,0 +1,166 @@
+        help="The instance to exec into.",
+        type=str,
+    ).completer = completer_clusters
+    pod_help = "The pod to exec into. Defaults to a random running pod for the service."
+    exec_parser.add_argument("-p", "--pod", help=pod_help)
+
+    exec_parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help=f"Define a different soa config directory. Defaults to %(default)s.",
+    )
+
+    exec_parser.set_defaults(command=paasta_exec)
+
+
+def completer_clusters(prefix, parsed_args, **kwargs):
+    service = parsed_args.service or guess_service_name()
+    if service in list_services():
+        return list_clusters(service)
+    else:
+        return list_clusters()
+
+
+def paasta_exec(args: argparse.Namespace) -> int:
+    """Execs into a running Paasta service.
+    :param args: argparse.Namespace obj created from sys.args by cli"""
+    soa_dir = args.soa_dir
+
+    service = figure_out_service_name(args, soa_dir)
+
+    if (
+        args.cluster is None
+        or args.instance is None
+        or len(args.instance.split(",")) > 2
+    ):
+        print(
+            PaastaColors.red("You must specify one cluster and one instance."),
+            file=sys.stderr,
+        )
+        return 1
+    cluster = args.cluster
+    if verify_instances(args.instance, service, cluster):
+        return 1
+
+    cluster = cluster[0]
+    instance = args.instance
+
+    if args.pod is None:
+        pod = (
+            subprocess.run(
+                [
+                    f"kubectl-{cluster}",
+                    "get",
+                    "po",
+                    "-n",
+                    "paasta",
+                    "|",
+                    "grep",
+                    "-E",
+                    f"'{service.replace('_','--')}-{instance.replace('_','--')}.*Running'",
+                    "|",
+                    "shuf",
+                    "|",
+                    "awk",
+                    "'END {{print $1}}'",
+                ],
+                stdout=subprocess.PIPE,
+            )
+            .stdout.decode("utf-8")
+            .strip()
+        )
+    else:
+        pods = (
+            subprocess.run(
+                [
+                    f"kubectl-{cluster}",
+                    "get",
+                    "po",
+                    "-n",
+                    "paasta",
+                    "|",
+                    "grep",
+                    "-E",
+                    f"'{service.replace('_','--')}-{instance.replace('_','--')}.*Running'",
+                    "|",
+                    "shuf",
+                    "|",
+                    "awk",
+                    "'{{print $1}}'",
+                ],
+                stdout=subprocess.PIPE,
+            )
+            .stdout.decode("utf-8")
+            .split("\n")
+        )
+        pod = args.pods
+        if pod not in pods:
+            pod = ""
```

Comment:
If we're checking if the pod they ask for is in the list of running pods, you can probably split out fetching the list of pods so you don't have to duplicate the code.